File: python-Raytracer/color.py
import math


def linear_to_gamma(linear_component):
    """Convert linear to gamma (gamma 2)"""
    return math.sqrt(linear_component) if linear_component > 0.0 else 0.0


def clamp(x, min_val, max_val):
    """Clamp x to the range [min_val, max_val]"""
    if x < min_val:
        return min_val
    elif x > max_val:
        return max_val
    else:
        return x


def write_color(file, pixel_color):
    """Write the color to the given file stream"""
    # Get components
    r = pixel_color.x()
    g = pixel_color.y()
    b = pixel_color.z()

    # Apply gamma correction
    r = linear_to_gamma(r)
    g = linear_to_gamma(g)
    b = linear_to_gamma(b)

    # Translate to [0,255] range
    intensity_min = 0.000
    intensity_max = 0.999
    r_byte = int(256 * clamp(r, intensity_min, intensity_max))
    g_byte = int(256 * clamp(g, intensity_min, intensity_max))
    b_byte = int(256 * clamp(b, intensity_min, intensity_max))

    # Write the bytes
    file.write(f"{r_byte} {g_byte} {b_byte}\n")

File: rust-Raytracer/src/color.rs
use std::io::Write;

use crate::vec3::Vec3;

pub type Color = Vec3;

fn linear_to_gamma(x: f64) -> f64 { if x > 0.0 { x.sqrt() } else { 0.0 } }

pub fn write_color<W: Write>(out: &mut W, pixel_color: Color) -> std::io::Result<()> {
    let mut r = pixel_color.0[0];
    let mut g = pixel_color.0[1];
    let mut b = pixel_color.0[2];

    r = linear_to_gamma(r);
    g = linear_to_gamma(g);
    b = linear_to_gamma(b);

    let clamp = |x: f64, min: f64, max: f64| if x < min { min } else if x > max { max } else { x };
    let intensity = (0.0, 0.999);
    let r_byte = (256.0 * clamp(r, intensity.0, intensity.1)) as i32;
    let g_byte = (256.0 * clamp(g, intensity.0, intensity.1)) as i32;
    let b_byte = (256.0 * clamp(b, intensity.0, intensity.1)) as i32;

    writeln!(out, "{} {} {}", r_byte, g_byte, b_byte)
}


